6312651: Compiler should only use verified interface types for optimization

rwestrel opened the pull request:
This change is mostly the same I sent for review 3 years ago but was
never integrated:

https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2019-May/033803.html

The main difference is that, in the meantime, I submitted a couple of
refactoring changes extracted from the 2019 patch:

8266550: C2: mirror TypeOopPtr/TypeInstPtr/TypeAryPtr with TypeKlassPtr/TypeInstKlassPtr/TypeAryKlassPtr
8275201: C2: hide klass() accessor from TypeOopPtr and typeKlassPtr subclasses

As a result, the current patch is much smaller (but still not small).

The implementation is otherwise largely the same as in the 2019
patch. I tried to remove some of the code duplication between the
TypeOopPtr and TypeKlassPtr hierarchies by having some of the logic
shared in template methods. In the 2019 patch, interfaces were trusted
when types were constructed and I had added code to drop interfaces
from a type where they couldn't be trusted. This new patch proceeds
the other way around: interfaces are not trusted when a type is
constructed and code that uses the type must explicitly request that
they are included (this was suggested as an improvement by Vladimir
Ivanov I think).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6312651](https://bugs.openjdk.org/browse/JDK-6312651): Compiler should only use verified interface types for optimization


### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [f49a042a](https://git.openjdk.org/jdk/pull/10901/files/f49a042ae3e9bb0d24648077258b94fc8e9c0e79)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10901/head:pull/10901` \
`$ git checkout pull/10901`

Update a local copy of the PR: \
`$ git checkout pull/10901` \
`$ git pull https://git.openjdk.org/jdk pull/10901/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10901`

View PR using the GUI difftool: \
`$ git pr show -t 10901`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10901.diff">https://git.openjdk.org/jdk/pull/10901.diff</a>

</details>

iwanowww commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -4725,7 +4836,8 @@ const Type *TypeAryPtr::xmeet_helper(const Type *t) const {
       }
-      return TypeInstPtr::make(ptr, ciEnv::current()->Object_klass(), false, NULL, offset, instance_id, speculative, depth);
+      interfaces = this_interfaces.intersection_with(tp_interfaces);
+      return TypeInstPtr::make(ptr, ciEnv::current()->Object_klass(), interfaces, false, NULL,offset, instance_id, speculative, depth);
```
> NULL,offset

missing space

iwanowww commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -5591,19 +5724,23 @@ const Type    *TypeInstKlassPtr::xmeet( const Type *t ) const {
       // below the centerline when the superclass is exact. We need to
       // do the same here.
-      if (klass()->equals(ciEnv::current()->Object_klass()) && !klass_is_exact()) {
+      if (klass()->equals(ciEnv::current()->Object_klass()) && this_interfaces.intersection_with(tp_interfaces).eq(this_interfaces) && !klass_is_exact()) {
```
> this_interfaces.intersection_with(tp_interfaces).eq(this_interfaces)

Maybe a case for a helper method `InterfaceSet::contains(InterfaceSet)`?

rwestrel replied:
Indeed, there are several uses of this pattern. I will make that change.

iwanowww commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -5637,81 +5775,117 @@ const Type    *TypeInstKlassPtr::xmeet( const Type *t ) const {
+    ciInstanceKlass* ik = k->as_instance_klass();
+    bool klass_is_exact = ik->is_final();
+    if (!klass_is_exact &&
+        deps != NULL && UseUniqueSubclasses) {
```
Please, put `UseUniqueSubclasses` guard at the top of the method.

rwestrel replied:
Ok.

iwanowww reviewed (commented):
Thanks, Roland! Overall, looks very good.

Submitted hs-tier1 - hs-tier4 testing. (Earlier, it went through hs-tier1 - hs-tier8 without new failures.)

Some minor comments/suggestions follow.

iwanowww commented:
BTW will there still be a need in `CastPP` vs `CheckCastPP` dichotomy after the patch goes in?

rwestrel commented on src/hotspot/share/ci/ciObjectFactory.cpp:
```diff
@@ -153,9 +153,16 @@ void ciObjectFactory::init_shared_objects() {
+  if (vmClasses::name##_is_loaded()) {                               \
+    InstanceKlass* ik = vmClasses::name();                           \
+    ciEnv::_##name = get_metadata(ik)->as_instance_klass();          \
+    Array<InstanceKlass*>* interfaces = ik->transitive_interfaces(); \
```
ciInstanceKlass objects for the vm classes all need to be allocated from the same long lived arena that's created in ciObjectFactory::initialize() because they are shared between compilations. Without that code, the ciInstanceKlass for a particular vm class is in the long lived arena but the ciInstanceKlass objects for the interfaces are created later when they are needed in the arena of some compilation. Once that compilation is over the interface objects are destroyed but still referenced from shared types such as TypeInstPtr::MIRROR.

rwestrel commented on src/hotspot/share/ci/ciInstanceKlass.cpp:
```diff
@@ -729,6 +729,19 @@ void ciInstanceKlass::dump_replay_instanceKlass(outputStream* out, InstanceKlass
+GrowableArray<ciInstanceKlass*>* ciInstanceKlass::transitive_interfaces() const {
+  GrowableArray<ciInstanceKlass*>* result = NULL;
+  GUARDED_VM_ENTRY(
+    InstanceKlass* ik = get_instanceKlass();
```
Maybe but it's not as simple as it seems (for the reason discussed below for vm classes). Some classes are allocated in a special arena because they are shared between compilations. A _transitive_interfaces array would need to be allocated in that same arena (otherwise if lazily allocated during a particular compilation it would be in the thread's resource are and  its content would be wiped out when the compilation is over but still reachable). That's not straightforward with the current implementation as the arena a ciInstanceKlass is allocated in is not available to the ciInstanceKlass so that would require extra changes and I'm not sure the extra complexity is worth it.

rwestrel replied:
scratch that. I see it's done for `_nonstatic_fields` and will do the same.

rwestrel commented on src/hotspot/share/opto/type.hpp:
```diff
@@ -1111,18 +1151,18 @@ class TypeOopPtr : public TypePtr {
   // If the klass is final, the resulting type will be exact.
-  static const TypeOopPtr* make_from_klass(ciKlass* klass) {
-    return make_from_klass_common(klass, true, false);
+  static const TypeOopPtr* make_from_klass(ciKlass* klass, bool trust_interface = false) {
```
Good suggestion. I will make that change.

rwestrel commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -566,6 +568,13 @@ void Type::Initialize_shared(Compile* current) {
   mreg2type[Op_RegFlags] = TypeInt::CC;
 
+  TypeAryPtr::_array_interfaces = new TypePtr::InterfaceSet();
+  GrowableArray<ciInstanceKlass*>* array_interfaces = ciArrayKlass::interfaces();
```
Good suggestion as well.

rwestrel commented:
> Thanks, Roland! Overall, looks very good.

Thanks for reviewing this. I pushed new commits that should address your comments/suggestions.

iwanowww commented on src/hotspot/share/ci/ciArrayKlass.cpp:
```diff
@@ -103,3 +104,17 @@ ciArrayKlass* ciArrayKlass::make(ciType* element_type) {
   }
 }
+
+GrowableArray<ciInstanceKlass*>* ciArrayKlass::interfaces() {
```
FTR there's a subtle asymmetry between `ciArrayKlass::interfaces()` and `ciInstanceKlass::transitive_interfaces()` when it comes to memory allocation: the former allocates from resource area while the latter from compiler arena.

It doesn't cause any problems since `ciArrayKlass::interfaces()` is used only in `Type::Initialize_shared()` to instantiate shared `TypeAryPtr::_array_interfaces`, but it took me some time to find that out. 

Maybe a helper method in `type.cpp` is a better place for that logic.

rwestrel replied:
How would that work? ciArrayKlass::interfaces() has to transition into the vm which is not something that would feel right in type.cpp.

iwanowww replied:
Indeed, good point.

IMO you could just work directly with CI mirrors for `Serializable` and `Cloneable`, but now I'm curious why does CI code diverge from `ObjArrayKlass::compute_secondary_supers()`?
  https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/objArrayKlass.cpp#L376

iwanowww replied:
And to answer my question: only `Serializable` and `Cloneable` are interfaces among those. The rest are arrays.

iwanowww commented on src/hotspot/share/ci/ciInstanceKlass.cpp:
```diff
@@ -729,6 +730,26 @@ void ciInstanceKlass::dump_replay_instanceKlass(outputStream* out, InstanceKlass
+GrowableArray<ciInstanceKlass*>* ciInstanceKlass::transitive_interfaces() {
+  if (_transitive_interfaces == NULL) {
+    GUARDED_VM_ENTRY(
+            InstanceKlass* ik = get_instanceKlass();
```
A candidate for `compute_transitive_interfaces()` helper method?

rwestrel replied:
Done in new commit.

iwanowww commented on src/hotspot/share/opto/subnode.cpp:
```diff
@@ -1047,7 +1047,7 @@ static inline Node* isa_const_java_mirror(PhaseGVN* phase, Node* n) {
   // return the ConP(Foo.klass)
   assert(mirror_type->is_klass(), "mirror_type should represent a Klass*");
-  return phase->makecon(TypeKlassPtr::make(mirror_type->as_klass()));
+  return phase->makecon(TypeKlassPtr::make(mirror_type->as_klass(),  Type::trust_interfaces));
```
Extra space.

rwestrel replied:
Done.

iwanowww commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -3767,62 +3962,62 @@ const TypeInstPtr* TypeInstPtr::cast_to_exactness(bool klass_is_exact) const {
+    //
+    assert(loaded->ptr() != TypePtr::Null, "insanity check");
+    //
+    if(      loaded->ptr() == TypePtr::TopPTR ) { return unloaded; }
```
Missing space after `if`.

rwestrel replied:
Done.

iwanowww commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -3767,62 +3962,62 @@ const TypeInstPtr* TypeInstPtr::cast_to_exactness(bool klass_is_exact) const {
+
+  // Both are unloaded, not the same class, not Object
+  // Or meet unloaded with a different loaded class, not java/lang/Object
+  if( ptr != TypePtr::BotPTR ) {
```
Missing space after `if`.

rwestrel replied:
Done.

iwanowww reviewed (commented):
Much better, thanks!

Minor comments/suggestions follow.

FTR test results are clean. I'll submit performance testing shortly.

rwestrel commented:
> Much better, thanks!
> 
> Minor comments/suggestions follow.
> 
> FTR test results are clean. I'll submit performance testing shortly.

Thanks. I updated the patch and removed the interface specific code in the VM_CLASS_DEFN() macro that's indeed no longer needed.

iwanowww commented:
FTR performance results look good.

iwanowww reviewed (approved):
Great work, Roland! I'm approving the PR. (hs-tier1 - hs-tier2 sanity testing passed with latest version.)

Feel free to handle `ciArrayKlass::interfaces()` as you find most appropriate.

rwestrel commented:
> Great work, Roland! I'm approving the PR. (hs-tier1 - hs-tier2 sanity testing passed with latest version.)
> 
> Feel free to handle `ciArrayKlass::interfaces()` as you find most appropriate.

Thanks for the review (and running tests).
I removed `ciArrayKlass::interfaces()` in the new commit. Is the resulting code what you suggested?

iwanowww commented:
> I removed ciArrayKlass::interfaces() in the new commit. Is the resulting code what you suggested?

Yes, it is much clearer now. Thanks.

vnkozlov reviewed (approved):
Nice work. Thanks!

rwestrel commented:
/integrate
